metalware-new/src/templater.rb

```diff
@@ -29,8 +29,25 @@
 require "constants"
 require 'nodeattr_interface'
 require 'metal_log'
+require 'exceptions'
 
 module Metalware
+  class TemplaterRecursiveOpenStruct < RecursiveOpenStruct
+    def initialize(*a)
+      @missing_tags = []
+      super(*a)
+    end
+
+    def method_missing(s, *a, &b)
+      value = super
+      if value.nil? && ! @missing_tags.include?(s)
+        @missing_tags.push s
+        MetalLog.warn "Missing template parameter: #{s}"
+      end
+      value
+    end
+  end
```

Comment:
I like the overall approach you've used for handling the missing parameters, I think it's simple and works well. One improvement I think we could make would be to use a separate object for doing this, something like this (untested, just an idea):

```ruby
class MissingParametersHandler
  def initialize(wrapped_object)
    @wrapped_object = wrapped_object
  end

  def method_missing(*a)
    @wrapped_object.send(a)
    # ... handle misisng parameters as you're doing ...
  end
end
```

We can then use this object to wrap both the `RecursiveOpenStruct` and the `MagicNamespace` object when these are created.

Advantages of this:

- we only need to have the logic of handling missing parameters in one place, rather than in the two places we have it at the moment; if we ever need to add this handling to additional objects in future we also only need to wrap them in a `MissingParametersHandler` rather than re-adding this behaviour to the new objects themselves;

- this separates out the handling of missing parameters from the other behaviour of the wrapped object, making it clearer why we're doing this;

- it avoids the use of inheritance when handling missing parameters for the `RecursiveOpenStruct` - in general you want to favour [composition over inheritance](https://en.wikipedia.org/wiki/Composition_over_inheritance#Benefits) unless you have a good reason not to; in particular you also normally want to avoid inheriting from classes you don't own which are not particularly intended to be inherited from, otherwise your code is more likely to break if these are changed in future.